### PR TITLE
session: return correct LastInsertID.

### PR DESCRIPTION
**Pull request by coocood**
When insert into an auto_increment column with a given value, the last insert id write to client should be the value.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @hanfei1991 PTAL

File: executor/write.go [hunk 1]
```diff
@@ -880,6 +880,7 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struc
 			}
 			row[i].SetInt64(val)
 			if val != 0 {
+				e.ctx.GetSessionVars().InsertID = uint64(val)
```
Comment (hanfei1991) [inline executor/write.go]: If there are two autoinc column, which value should be choosed?

File: session.go [hunk 2]
```diff
@@ -159,7 +159,10 @@ func (s *session) Status() uint16 {
 }
 
 func (s *session) LastInsertID() uint64 {
-	return s.sessionVars.LastInsertID
+	if s.sessionVars.LastInsertID > 0 {
```
Comment (hanfei1991) [inline session.go]: When LastInsertID is 0?

File: session_test.go [hunk 3]
```diff
@@ -521,11 +540,11 @@ func (s *testSessionSuite) TestIssue827(c *C) {
 	r, err := GetRows(rs)
 	c.Assert(err, IsNil)
 	c.Assert(r, DeepEquals, expect)
-	currLastInsertID := se.LastInsertID()
+	currLastInsertID := se.GetSessionVars().PrevLastInsertID
```
Comment (hanfei1991) [inline session_test.go]: Why change these?

File: executor/write.go (hunk 1 above)
Comment (coocood) [inline executor/write.go]: there can be only one auto column in a table.

File: session.go (hunk 2 above)
Comment (coocood) [inline session.go]: It is set to zero when we reset statement context.

File: session_test.go (hunk 3 above)
Comment (coocood) [inline session_test.go]: LastInsertID is reset to zero on every request, its value is set to PrevLastInsertID

**Comment by hanfei1991** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTm

File: session_test.go [hunk 4]
```diff
@@ -277,13 +277,32 @@ func (s *testSessionSuite) TestAutoIncrementID(c *C) {
 	c.Assert(lastID, Less, uint64(4))
 	mustExecSQL(c, se, "insert t () values ()")
 	c.Assert(se.LastInsertID(), Greater, lastID)
-	mustExecSQL(c, se, "insert t () select 100")
+	lastID = se.LastInsertID()
+	mustExecSQL(c, se, "insert t values (100)")
+	c.Assert(se.LastInsertID(), Equals, uint64(100))
+
+	// If the auto_increment column value is given, it uses the value of the latest row.
+	mustExecSQL(c, se, "insert t values (120), (112)")
```
Comment (zimulala) [inline session_test.go]: How about the case of `insert t values (120), (112),()`?

Comment (zimulala) [inline session_test.go]: I got it.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**